### ISSUE TITLE
[Bug] Keys of ingested file are lost

### ISSUE THREAD

**Issue by Little-Wallace** [facebook/rocksdb#6666]
> Note: Please use Issues only for bug reports. For questions, discussions, feature requests, etc. post to dev group: https://groups.google.com/forum/#!forum/rocksdb or https://www.facebook.com/groups/rocksdb.dev

We  queried the keys in which files we ingested just before and put them into another `SstFileWriter`, then we got error "Keys must be added in strict ascending order.".

We have found out the reason why keys are unordered. This is because RocksDB will encode data block and make keys with the same prefix can share prefix string. This feature makes RocksDB save much space when there are many similar keys. RocksDB encodes these keys with `InternalKey` rather than `UserKey`. If a key is very short and after it is coded as InternalKey, the common prefix with his adjacent key is longer than itself (we call it UserKey), BlockBasedTable will reset the seqno of this key if it exists in an ingested file and **make the common prefix changed incorrectly.**
 You can reproduce the bug by this test: [Little-Wallace/rocksdb](https://github.com/Little-Wallace/rocksdb/blob/out-of-order/db/external_sst_file_basic_test.cc#L1153)
This bug is caused by https://github.com/facebook/rocksdb/blob/master/table/block_based/block.cc#L555

**Comment by Little-Wallace** [comment]
Here is test code which reproduces this problem.
```
  Options options = CurrentOptions();
  options.force_consistency_checks = true;

  DestroyAndReopen(options);
  const size_t kValueSize = 1 << 3;
  Random rnd(301);
  std::string value(RandomString(&rnd, kValueSize));

  ExternalSstFileInfo info;
  std::string f = test::PerThreadDBPath("sst_file_0");
  EnvOptions env;
  rocksdb::SstFileWriter writer(env, options);
  // Write some key to make global seqno larger than zero
  for (int i = 0; i < 10; i++) {
    ASSERT_OK(Put("ab" + Key(i), value));
  }
  // Get a Snapshot to make RocksDB assign global seqno to ingested sst files.
  auto snap = dbfull()->GetSnapshot();

  auto s = writer.Open(f);
  ASSERT_OK(s);
  std::string key1 = "ab";
  std::string key2 = "ab";

  // Make the prefix of key2 is same with key1 add zero seqno. The tail of every
  // key is composed as (seqno << 8 | value_type), and here `1` represents
  // ValueType::kTypeValue
  PutFixed64(&key2, 1);
  key2 += "cdefghijkl";

  ASSERT_OK(writer.Put(key1, value));
  ASSERT_OK(writer.Put(key2, value));

  ASSERT_OK(writer.Finish(&info));
  IngestExternalFileOptions ingest_opt;
  ingest_opt.move_files = true;

  ASSERT_OK(dbfull()->IngestExternalFile({info.file_path}, ingest_opt));
  dbfull()->ReleaseSnapshot(snap);
  ASSERT_EQ(value, Get(key1));
  // You will get error here
  ASSERT_EQ(value, Get(key2));
```

**Comment by zhangjinpeng1987** [comment]
@Little-Wallace Can we fix it?

**Comment by mm304321141** [comment]
try add `BlockIter::key_footer_backup_` to save the real seq type ... ?
`DataBlockIter::ParseNextDataKey` restore the `key_footer_backup_ ` before `key_.TrimAppend(shared, p, non_shared);` ?
or ...
 add `BlockIter::key_backup_` to save the key before `key_.UpdateInternalKey(global_seqno_, value_type);`

**Comment by ajkr** [comment]
Thanks for the issue. It's a great finding!

**Comment by yiwu-arbug** [comment]
Can anyone take a look and suggest a fix? cc @siying

**Comment by ajkr** [comment]
Fixed by #6669.
